fplll/pruner/pruner.h

```diff
@@ -290,47 +390,207 @@ template <class FT> class Pruner
-
     set_tabulated_consts();
+    // need to fix target if possible
+    if (metric == PRUNER_METRIC_PROBABILITY_OF_SHORTEST)
+    {
+      // if target > 1 or < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (this->target > 1.0 || this->target < 0.0)
+      {
+        this->target = 0.99;
+        opt_overall  = true;
+      }
+    }
+    else if (metric == PRUNER_METRIC_EXPECTED_SOLUTIONS)
+    {
+      // if target < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (this->target < 0.0)
+      {
+        this->target = 0.99;
+        opt_overall  = true;
+      }
+    }
+    else
+    {
+      throw std::invalid_argument("Pruner was set to an unknown metric");
+    }
     load_basis_shapes(gso_rs);
   }
 
   /**
+     @brief run the optimization process using 'even' coefficients. Note
+     the optimization only applies to the pruning coefficients indexed
+     by (0, 2, 4, ... n). It thus uses half of the coefficients.
+
+     Run the optimization process, successively using the algorithm activated
+     using using half coefficients.
+  */
+  void optimize_coefficients_evec(/*io*/ vector<double> &pr);
+
+  /**
+     @brief run the optimization process using all the coefficients.
+
      Run the optimization process, successively using the algorithm activated
+     using using full coefficients.
 
-     @brief run the optimization process
   */
-  void optimize_coefficients(/*io*/ vector<double> &pr);
+  void optimize_coefficients_full(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to reduce single enumeration time
+
+     Optimization process to the pruning parameters to reduce single enumeration
+     time with the hope that it also reduces the overall enumeration time.
+  */
+  void optimize_coefficients_tune_cost(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to increase succ. probability
+
+     Optimization process to the pruning parameters to increase succ.
+     probability with the restriction that the single enumeration time does
+     not increase significantly.
+  */
+  void optimize_coefficients_tune_prob(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to make the curve a more smooth.
+
+     Optimization process to the pruning parameters to make the curve
+     more smooth if the input has a large gap between consecutive
+     pruning parameters.
+  */
+  void optimize_coefficients_smooth(/*io*/ vector<double> &pr);
+
+  /**
+     @brief main interface to optimize the overall enumeraiton time
+
+     Main interface to optimize the overall enumeraiton time where the
+     target unciton is: single_enumeration_cost divided by succ. probability
+     or expected solutions.
+  */
+  void optimize_coefficients_cost(/*io*/ vector<double> &pr);
```

Comment:
This doesn't seem correct, the overall cost is (t-1)*preproc + t*enum where t is roughly 1/pr, no?

Reply:
Yes, fixing now.